src/main/scala/com/databricks/labs/overwatch/utils/Upgrade.scala

```diff
@@ -15,15 +17,31 @@ object Upgrade extends SparkSessionWrapper {
 
   import spark.implicits._
 
+  private def persistPipelineStateChange(db: String, moduleIds: Array[Int], asOf: Option[java.sql.Timestamp] = None): Unit = {
+    val logMsg = s"ROLLING BACK STATE: Modules ${moduleIds.mkString(",")} are rolling back"
+    println(logMsg)
+    logger.log(Level.INFO, logMsg)
+
+    val baseUpdateSQL = s"update ${db}.pipeline_report set status = 'UPGRADED' " +
+      s"where moduleID in (${moduleIds.mkString(", ")}) " +
+      s"and (status = 'SUCCESS' or status like 'EMPT%') "
+    val updateStatement = if (asOf.nonEmpty) {
+      baseUpdateSQL + s"and Pipeline_SnapTS > '${asOf.get.toString}' "
+    } else baseUpdateSQL
+
+    spark.sql(updateStatement)
+  }
+
+  private def getNumericalSchemaVersion(version: String): Int = {
+    version.split("\\.").reverse.head.toInt
+  }
+
   @throws(classOf[UpgradeException])
   private def validateSchemaUpgradeEligibility(currentVersion: String, targetVersion: String): Unit = {
     var valid = true
-    val currentVersionDotArray = currentVersion.split("\\.")
-    val targetVersionDotArray = targetVersion.split("\\.")
-    currentVersionDotArray.dropRight(1).zipWithIndex.foreach(currentV => {
-      valid = currentV._1 == targetVersionDotArray(currentV._2)
-    })
-    valid = valid && currentVersionDotArray.reverse.head.toInt < targetVersionDotArray.reverse.head.toInt
+    val currentNumericalVersion = getNumericalSchemaVersion(currentVersion)
+    val targetNumericalVersion = getNumericalSchemaVersion(targetVersion)
+    valid = targetNumericalVersion > currentNumericalVersion
```

Comment:
if we're supporting major.minor, then we'll need to check major part as well

Reply:
I would better introduce the Version class, and do correct major.minor.patch comparison - it's easy...